File: baeckjoon/2178.java
import java.io.*;
import java.util.*;

class Main{

    static int N, M;
    
    static int[][] maze;
    static boolean[][] visited;

    static int dx[] = {0,0,-1,1};
    static int dy[] = {1,-1,0,0};

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        N = Integer.parseInt(st.nextToken());
        M = Integer.parseInt(st.nextToken());

        maze = new int[N][M];
        visited = new boolean[N][M];

        for (int i = 0; i < N; i++) {
            st = new StringTokenizer(br.readLine());
            String line = st.nextToken();
            for (int j = 0; j < M; j++) {
                maze[i][j] = line.charAt(j) - '0';
            }
        }

        bfs(0,0);

        /*
        // maze check
        for (int i = 0; i < N; i++){
            for (int j = 0; j < M; j++){
                System.out.printf("%d ",maze[i][j]);
            }
            System.out.println();
        }
        */
        
        System.out.println(maze[N-1][M-1]);
    }

    public static void bfs(int i, int j){
        Queue<int[]> q = new LinkedList<>();
        q.offer(new int[] {i,j});

        while(!q.isEmpty()){
            int location[] = q.poll();
            visited[i][j] = true;

            for (int dir = 0; dir < 4; dir++){
                int x = location[0] + dx[dir];
                int y = location[1] + dy[dir];

                if (x >= 0 && y >= 0 && x < N && y < M){
                    if (maze[x][y] != 0 && !visited[x][y]){
                        q.offer(new int[] {x,y});
                        visited[x][y] = true;
                        maze[x][y] = maze[location[0]][location[1]] + 1;
                    }
                }
            }
        }
    }
}

File: baeckjoon/2178.py
from collections import deque
# N, M 입력
N, M = map(int,input().split())
# 미로 초기화
maze = []
for i in range(N):
    maze.append(list(map(int,list(input()))))
# 큐, 방문 여부
q = deque([[0,0]])
visited = [[False for j in range(M)] for i in range(N)]
# BFS
while q:
    # 현재 위치
    location = q.popleft()
    for di, dj in zip([-1,1,0,0],[0,0,-1,1]):
        # 이동할 좌표
        i, j = location[0] + di, location[1] + dj
        # 좌표 유효성 검증
        if i < 0 or j < 0 or i >= N or j >= M:
            continue
        # 방문 여부 확인 및 이동 가능 여부 확인
        if not visited[i][j] and maze[i][j] != 0:
            q.append([i,j])
            maze[i][j] = maze[location[0]][location[1]] + 1
            visited[i][j] = True

print(maze[N-1][M-1])
